Allow text values to specify a language

SepidehAlassi opened the pull request:
- [x] Create text values with language tags in API v1 via bulk import.
- [x] Create and update text values with language tags in API v1 via the `/v1/resources` and `/v1/values` routes.
- [x] Filter text values by language in KnarQL in API v2.
- [x] Add tests.
  - [x] Add bulk import tests.
  - [x] Add other tests.
- [x] Update documentation.
  - [x] Update bulk import docs.
  - [x] Update other API docs.
- [x] Update release notes.

Requires a change in SALSAH 2, which is implemented in https://github.com/dhlab-basel/Salsah/pull/247.

Resolves #736.

tobiasschweizer commented on webapi/src/main/scala/org/knora/webapi/util/standoff/StandoffTagUtilV2.scala:
```diff
@@ -195,7 +195,7 @@ object StandoffTagUtilV2 {
       * @param text          the text as a mere sequence of characters.
       * @param standoffTagV2 the text's standoff markup.
```
Could you add the documentation for the language param?

tobiasschweizer commented on webapi/src/main/twirl/queries/sparql/v1/generateInsertStatementsForCreateValue.scala.txt:
```diff
@@ -59,15 +59,21 @@
         @value match {
 
             case textValueSimple: TextValueSimpleV1 => {
```
I think this logic has to be added to `src/main/twirl/queries/sparql/v1/addValueVersion.scala.txt` too.

tobiasschweizer replied:
Or can the language only be added when creating new values using the bulk import?

SepidehAlassi replied:
no, it is not supposed to work just with bulk import. This logic is currently implemented for creation of TextValue for first time both for bulk import requests and individual resource creation requests. You are right, the same logic can be added in case a TextValue already exists and a new version of it should be added with specified language.

tobiasschweizer replied:
So some logic needs to be added to

- `ResourcesResponderV1`
- `ValuesResponderV1`

I can help you with that if necessary.

SepidehAlassi commented:
@benjamingeer is there any hard-coded ontology for simple schema somewhere?
when I try full resource request for text values with language, it works for complex schema but fails for simple schema, with "unknown key error"
To reproduce the issue you can add following test in ResourcesRouteV2R2RSpec :
`  "perform a full resource request for a resource with a Text language (in the simple schema)" in {

            Get(s"/v2/resources/${URLEncoder.encode("http://rdfh.ch/0001/a-thing-with-text-valuesLanguage", "UTF-8")}").addHeader(new SchemaHeader(RouteUtilV2.SIMPLE_SCHEMA_NAME)) ~> resourcesPath ~> check {

                assert(status == StatusCodes.OK, response.toString)
            }
        }`

Error message is:

[2018-05-14 17:46:41,842] ERROR - ResourcesRouteV2R2Spec - Unable to run route /v2/resources/http:%2F%2Frdfh.ch%2F0001%2Fa-thing-with-text-valuesLanguage
com.github.jsonldjava.core.JsonLdError: invalid value object: value object has unknown keys
	at com.github.jsonldjava.core.JsonLdApi.expand(JsonLdApi.java:928)
	at com.github.jsonldjava.core.JsonLdApi.expand(JsonLdApi.java:522)
	at com.github.jsonldjava.core.JsonLdApi.expand(JsonLdApi.java:847)
	at com.github.jsonldjava.core.JsonLdApi.expand(JsonLdApi.java:522)
	at com.github.jsonldjava.core.JsonLdApi.expand(JsonLdApi.java:847)
	at com.github.jsonldjava.core.JsonLdApi.expand(JsonLdApi.java:1025)
	at com.github.jsonldjava.core.JsonLdProcessor.expand(JsonLdProcessor.java:149)
	at com.github.jsonldjava.core.JsonLdProcessor.compact(JsonLdProcessor.java:52)
	at org.knora.webapi.util.jsonld.JsonLDDocument.toPrettyString(JsonLDUtil.scala:396)
	at org.knora.webapi.routing.RouteUtilV2$.$anonfun$runJsonRoute$2(RouteUtilV2.scala:133)
	at scala.util.Success.$anonfun$map$1(Try.scala:251)
	at scala.util.Success.map(Try.scala:209)
	at scala.concurrent.Future.$anonfun$map$1(Future.scala:289)
	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:29)
	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:29)
	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
	at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
	at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
	at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
	at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
	at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:43)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)

benjamingeer commented:
> is there any hard-coded simple ontology schema somewhere?

The `knora-api` ontologies (simple and complex) aren’t hard-coded anymore, they’re generated on the fly from `knora-base`, following the rules in the classes `KnoraApiV2Simple` and `KnoraApiV2WithValueObjects`.

You can see them as JSON-LD in the test data:

https://github.com/dhlab-basel/Knora/blob/develop/webapi/src/test/resources/test-data/ontologyR2RV2/knoraApiOntologySimple.json
https://github.com/dhlab-basel/Knora/blob/develop/webapi/src/test/resources/test-data/ontologyR2RV2/knoraApiWithValueObjects.json

(I think those files have changes on this branch.)

I’ll look at the error you’re getting tomorrow.

benjamingeer commented:
@SepidehAlassi That error is my mistake, fixing it now...

benjamingeer commented on webapi/src/test/resources/test-data/searchR2RV2/bookWithIncomingPagesOnlyLink.jsonld:
```diff
@@ -17,11 +17,13 @@
           "@type" : "knora-api:LinkValue",
-          "knora-api:linkValueHasTargetIri" : "http://rdfh.ch/8be1b7cf7103"
+          "knora-api:linkValueHasTargetIri" : {
+            "@id" : "http://rdfh.ch/8be1b7cf7103"
```
@tobiasschweizer When I changed the way IRIs are represented in JSON-LD, I missed `linkValueHasSourceIri` and `linkValueHasTargetIri`. It's fixed here. Does this require a change in SALSAH?

tobiasschweizer replied:
Yes, I think so. 

`src/app/model/webapi/knora/v2/convert-jsonld.ts`

```typescript
case AppConfig.LinkValue:

                let linkValue: ReadLinkValue;

                // check if the referred resource is given as an object or just as an IRI
                if (propValue[AppConfig.linkValueHasTarget] !== undefined) {
                    // linkValueHasTarget contains the object

                    let referredResource: ReadResource = constructReadResource(propValue[AppConfig.linkValueHasTarget]);

                    linkValue = new ReadLinkValue(propValue['@id'], propIri, referredResource.id, referredResource);
                } else if (propValue[AppConfig.linkValueHasTargetIri] !== undefined) {
                    // linkValueHasTargetIri contains the resource's Iri

                    let referredResourceIri = propValue[AppConfig.linkValueHasTargetIri];

                    linkValue = new ReadLinkValue(propValue['@id'], propIri, referredResourceIri);
                } else if (propValue[AppConfig.linkValueHasSource] !== undefined) {
                    // linkValueHasSource contains the object

                    let incomingResource: ReadResource = constructReadResource(propValue[AppConfig.linkValueHasSource]);

                    linkValue = new ReadLinkValue(propValue['@id'], propIri, incomingResource.id, incomingResource);
                } else if (propValue[AppConfig.linkValueHasSourceIri] !== undefined) {
                    // linkValueHasSourceIri contains the resource's Iri

                    let incomingResourceIri = propValue[AppConfig.linkValueHasSourceIri];

                    linkValue = new ReadLinkValue(propValue['@id'], propIri, incomingResourceIri);
                }

                valueSpecificProp = linkValue;
                break;
```

This is an object too now, right? So  we need to add `['@id']`. A test should be added to Salsah (on branch `unit tests`).

I can try to do it this afternoon.

tobiasschweizer replied:
Do you have a test response in JDON-LD complex I could use as a test case?

I think the Iris are only returned in complex schema if there are circular references.

benjamingeer replied:
Thanks, sorry about that!

The test responses are:

* `webapi/src/test/resources/test-data/searchR2RV2/bookWithIncomingPagesOnlyLink.jsonld`
* `webapi/src/test/resources/test-data/searchR2RV2/bookWithIncomingPagesWithAllRequestedProps.jsonld`

tobiasschweizer replied:
done in https://github.com/dhlab-basel/Salsah/pull/247/commits/074ae6a7a5366bffb7ee44a517ad5fe72b2dffef

benjamingeer replied:
Thank you!

benjamingeer commented on webapi/src/main/scala/org/knora/webapi/messages/v2/responder/resourcemessages/ResourceMessagesV2.scala:
```diff
@@ -84,31 +84,30 @@ case class ReadValueV2(valueIri: IRI, valueContent: ValueContentV2) extends IOVa
-                            ("@id" -> JsonLDString(valueIri)) +
-                            ("@type" -> JsonLDString(valueContent.valueType.toString))
-                    )
+        // In the complex schema, add the value's IRI and type to the JSON-LD object that represents it.
```
@SepidehAlassi This fixes the error you were getting, sorry about that.

SepidehAlassi replied:
no problem, thank you for quick fix!

SepidehAlassi commented:
@benjamingeer @tobiasschweizer I have added lots of tests and added textValue language option to api-doc. I also have added the functionality to create and update text values with language tags in API v1 via the /v1/resources and /v1/values routes, there are also tests for it. Can you please review these commits? Thanks.

benjamingeer commented on webapi/src/test/scala/org/knora/webapi/e2e/v2/SearchRouteV2R2RSpec.scala:
```diff
@@ -1861,10 +1863,61 @@ class SearchRouteV2R2RSpec extends R2RSpec {
+                  |
+                  |     FILTER(lang(?text) = "fr")
+                  |
+                  |     FILTER(?text = "Bonjour")
```
Does this work if you write:

```
FILTER(lang(?text) = "fr" && ?text = "Bonjour")
```

tobiasschweizer commented:
Yes, because it affects the JSON format.

SepidehAlassi commented:
Done!

tobiasschweizer commented:
language is added as optional prop :-)
